roxie/udplib/udptrr.cpp

```diff
@@ -433,20 +671,171 @@ class CReceiveManager : implements IReceiveManager, public CInterface
+                // the permit has timed out.
+                sender->setNumPermits(msg.packets);
+                break;
+
+            case flowType::send_completed:
+                noteDone(sender, msg);
+                break;
+
+            case flowType::request_to_send_more:
+                noteDone(sender, msg);
+                noteRequest(sender, flowSeq+1, msg.sendSeq);
+                break;
+
+            default:
+                DBGLOG("UdpReceiver: received unrecognized flow control message cmd=%i", msg.cmd);
+            }
         }
 
-        virtual void start()
+        unsigned checkPendingRequests()
         {
-            running = true;
-            Thread::start();
+            unsigned timeout = 5000;   // The default timeout is 5 seconds if nothing is waiting for response...
+            unsigned permitsIssued = 0;
+            if (pendingPermits)
+            {
+                unsigned now = msTick();
+                UdpSenderEntry *finger = pendingPermits;
+                while (finger)
+                {
+                    unsigned elapsed = now - finger->permitTime;
+                    if (elapsed >= udpPermitTimeout)
+                    {
+                        if (udpTraceLevel || udpTraceFlow || udpTraceTimeouts)
+                        {
+                            StringBuffer s;
+                            DBGLOG("permit %" SEQF "u to node %s (%u packets) timed out after %u ms, rescheduling", finger->flowSeq, finger->dest.getIpText(s).str(), finger->getActivePermits(), elapsed);
+                        }
+                        UdpSenderEntry *next = finger->nextSender;
+                        pendingPermits.remove(finger);
+                        if (++finger->timeouts > udpMaxRetryTimedoutReqs && udpMaxRetryTimedoutReqs != 0)
+                        {
+                            if (udpTraceLevel || udpTraceFlow || udpTraceTimeouts)
+                            {
+                                StringBuffer s;
+                                DBGLOG("permit to send %" SEQF "u to node %s timed out %u times - abandoning", finger->flowSeq, finger->dest.getIpText(s).str(), finger->timeouts);
+                            }
+                        }
+                        else
+                        {
+                            // Put it back on the queue (at the back)
+                            finger->permitTime = 0;
+                            pendingRequests.append(finger);
+                            finger->state = flowType::request_to_send;
+                        }
+                        finger = next;
+                    }
+                    else
+                    {
+                        timeout = udpPermitTimeout - elapsed;
+                        break;
+                    }
+                }
+
+                // Sum any remaining permits
+                while (finger)
+                {
+                    permitsIssued += finger->getActivePermits();
```

Comment:
Shouldn't we have accumulated these inside the loop above as well? I don't see where we do so.

Reply:
ok